### ISSUE TITLE
Geomopt Example

### ISSUE THREAD

**Issue by jordangarside** [pyscf/pyscf#147]
Geometry optimisation is working well by itself, but there is a bug in the [qmmm example](https://github.com/sunqm/pyscf/blob/c60ca2ee043e9a01b5269c48eefe4d8b6763ccfb/examples/geomopt/10-with_qmmm.py):
~~~py
/home/animal/Packages/pyberny/berny/Math.py:129: RuntimeWarning: divide by zero encountered in double_scalars
  x = x-fx/dxf
/home/animal/Packages/pyberny/berny/berny.py:182: RuntimeWarning: invalid value encountered in multiply
  return norm(np.linalg.solve(l*eye(H.shape[0])-H, g))-trust
Traceback (most recent call last):
  File "10-with_qmmm.py", line 24, in <module>
    mol1 = berny_solver.optimize(mf)
  File "/home/animal/Packages/pyscf/pyscf/geomopt/berny_solver.py", line 63, in optimize
    optimizer.send((energy, gradients))
  File "/home/animal/Packages/pyberny/berny/berny.py", line 36, in Berny
    converged = algo.step(energy, gradients, log=log)
  File "/home/animal/Packages/pyberny/berny/berny.py", line 94, in step
    dot(proj, s.interpolated.g), H_proj, s.weights, s.trust, log=log
  File "/home/animal/Packages/pyberny/berny/berny.py", line 183, in quadratic_step
    l = Math.findroot(steplength, ev[0])  # minimization on sphere
  File "/home/animal/Packages/pyberny/berny/Math.py", line 136, in findroot
    raise FindrootException()
berny.Math.FindrootException
~~~

**Comment by azag0** [comment]
I'll have a look at this

**Comment by azag0** [comment]
The real issue here is that the trust radius gets too small. This would indicate that the forces are incorrect. I've tried to optimize the same geometry with Mopac, and there is no issue.

Are we sure that the QM/MM gradients from pyscf are correct?

**Comment by sunqm** [comment]
The qmmm gradients has a bug and it is now fixed in adc120770.  The finite difference and analytical gradients of that example match well. The geometry optimization procedure still fails with berny.Math.FindrootException.

If you are on the dev branch, the code below can compute gradients with finite displacement and feed to berny solver
```
from pyscf.geomopt import berny_solver
def finite_diff_grad(mol):
    mfs = mf.as_scanner()
    mfs.verbose = 0
    e0 = mfs(mol)
    px = mol._atm[:,gto.PTR_COORD]
    py = mol._atm[:,gto.PTR_COORD] + 1
    pz = mol._atm[:,gto.PTR_COORD] + 2
    idx = numpy.hstack((px,py,pz))
    de = []
    for i in idx:
        mol._env[i] += 0.001
        e1 = mfs(mol)
        mol._env[i] -= 0.002
        e2 = mfs(mol)
        de.append((e1 - e2) / 0.002)
        mol._env[i] += 0.001
    de = numpy.array(de).reshape(3,-1).T
    print('E', e0, 'nuc_grad')
    print(de)
    return e0, de
m = berny_solver.as_pyscf_method(mol, finite_diff_grad)
m.verbose = 4
berny_solver.optimize(m)
```
